// @lit-protocol/auth-browser/src/lib/chains/sol.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.signAndSaveAuthMessage = exports.checkAndSignSolAuthMessage = exports.connectSolProvider = void 0;
const constants_1 = require("@lit-protocol/constants");
const misc_1 = require("@lit-protocol/misc");
const misc_browser_1 = require("@lit-protocol/misc-browser");
// import { toString as uint8arrayToString } from 'uint8arrays';
const uint8arrays_1 = require("@lit-protocol/uint8arrays");
/**
 *
 * Get the Solana provider from the browser web3 extension
 *
 * @returns { object || never }
 */
const getProvider = () => {
    let resultOrError;
    // -- validate
    if ('solana' in window) {
        resultOrError = (0, constants_1.ERight)(window?.keplr ?? window?.solana);
    }
    else {
        // -- finally
        const message = 'No web3 wallet was found that works with Solana.  Install a Solana wallet or choose another chain';
        resultOrError = (0, constants_1.ELeft)({
            message,
            error: constants_1.LIT_ERROR.NO_WALLET_EXCEPTION,
        });
    }
    return resultOrError;
};
/**
 *
 * Get Solana provider
 *
 * @returns { Promise<IProvider | undefined }
 */
const connectSolProvider = async () => {
    const providerOrError = getProvider();
    if (providerOrError.type === 'ERROR') {
        (0, misc_1.throwError)(providerOrError.result);
        return;
    }
    let provider = providerOrError.result;
    await provider.connect();
    const account = provider.publicKey.toBase58();
    return { provider, account };
};
exports.connectSolProvider = connectSolProvider;
/**
 *
 * Check and sign solana auth message
 *
 * @returns { JsonAuthSig }
 */
const checkAndSignSolAuthMessage = async () => {
    const res = await (0, exports.connectSolProvider)();
    if (!res) {
        (0, misc_1.log)('Failed to connect sol provider');
    }
    const provider = res?.provider;
    const account = res?.account;
    const key = constants_1.LOCAL_STORAGE_KEYS.AUTH_SOL_SIGNATURE;
    let authSigOrError = (0, misc_browser_1.getStorageItem)(key);
    // let authSig = localStorage.getItem("lit-auth-sol-signature");
    let authSig;
    // -- case: if unable to get auth from local storage
    if (authSigOrError.type === "ERROR" /* EITHER_TYPE.ERROR */) {
        (0, misc_1.log)('signing auth message because sig is not in local storage');
        await (0, exports.signAndSaveAuthMessage)({ provider });
        authSigOrError.type = "SUCCESS" /* EITHER_TYPE.SUCCESS */;
        authSigOrError.result = (0, misc_browser_1.getStorageItem)(key);
    }
    //   @ts-ignore
    window.test = authSigOrError;
    try {
        // when it's not in local storage, it's a string
        authSig = JSON.parse(authSigOrError.result.result);
    }
    catch (e) {
        // when it's in local storage, it's an object
        authSig = JSON.parse(authSigOrError.result);
    }
    // -- if the wallet address isn't the same as the address from local storage
    if (account !== authSig.address) {
        (0, misc_1.log)('signing auth message because account is not the same as the address in the auth sig');
        await (0, exports.signAndSaveAuthMessage)({ provider });
        authSigOrError.type = "SUCCESS" /* EITHER_TYPE.SUCCESS */;
        authSigOrError.result = (0, misc_browser_1.getStorageItem)(key);
        authSig = JSON.parse(authSigOrError.result);
    }
    (0, misc_1.log)('authSig', authSig);
    return authSig;
};
exports.checkAndSignSolAuthMessage = checkAndSignSolAuthMessage;
/**
 *
 * Sign and save auth signature locally (not saved to the nodes)
 *
 * @property { any } provider
 * @return { Promise<JsonAuthSig | undefined> }
 *
 */
const signAndSaveAuthMessage = async ({ provider, }) => {
    const now = new Date().toISOString();
    const body = constants_1.AUTH_SIGNATURE_BODY.replace('{{timestamp}}', now);
    //   turn body into Uint8Array
    const data = (0, uint8arrays_1.uint8arrayFromString)(body, 'utf8');
    //   const data = naclUtil.encode(body);
    const signed = await provider.signMessage(data, 'utf8');
    const hexSig = (0, uint8arrays_1.uint8arrayToString)(signed.signature, 'base16');
    const authSig = {
        sig: hexSig,
        derivedVia: 'solana.signMessage',
        signedMessage: body,
        address: provider.publicKey.toBase58(),
    };
    localStorage.setItem(constants_1.LOCAL_STORAGE_KEYS.AUTH_SOL_SIGNATURE, JSON.stringify(authSig));
    return authSig;
};
exports.signAndSaveAuthMessage = signAndSaveAuthMessage;
