// @lens-protocol/domain/entities/dist/lens-protocol-domain-entities.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var PublicationReport = require('../../dist/PublicationReport-0041056c.cjs.prod.js');
var Publication = require('../../dist/Publication-5ba726a5.cjs.prod.js');
var Transactions = require('../../dist/Transactions-61045839.cjs.prod.js');
var createClass = require('../../dist/createClass-df27ed77.cjs.prod.js');
var createSuper = require('../../dist/createSuper-671c6ebe.cjs.prod.js');
var wrapNativeSuper = require('../../dist/wrapNativeSuper-7caa6ff8.cjs.prod.js');
require('../../dist/unsupportedIterableToArray-afbea1dd.cjs.prod.js');

var NftContractType = /*#__PURE__*/function (NftContractType) {
  NftContractType["Erc721"] = "ERC721";
  NftContractType["Erc1155"] = "ERC1155";
  return NftContractType;
}({});

var ProfileReportReason = /*#__PURE__*/function (ProfileReportReason) {
  ProfileReportReason["IMPERSONATION"] = "IMPERSONATION";
  ProfileReportReason["FRAUD_SOMETHING_ELSE"] = "FRAUD_SOMETHING_ELSE";
  ProfileReportReason["REPETITIVE"] = "REPETITIVE";
  ProfileReportReason["SPAM_SOMETHING_ELSE"] = "SPAM_SOMETHING_ELSE";
  return ProfileReportReason;
}({});

var InsufficientGasError = /*#__PURE__*/function (_Error) {
  createSuper._inherits(InsufficientGasError, _Error);
  var _super = createSuper._createSuper(InsufficientGasError);
  function InsufficientGasError(asset) {
    var _this;
    createClass._classCallCheck(this, InsufficientGasError);
    _this = _super.call(this, 'Not enough gas to pay for the operation');
    wrapNativeSuper._defineProperty(createSuper._assertThisInitialized(_this), "name", 'InsufficientGasError');
    _this.asset = asset;
    return _this;
  }
  return createClass._createClass(InsufficientGasError);
}( /*#__PURE__*/wrapNativeSuper._wrapNativeSuper(Error));
var PendingSigningRequestError = /*#__PURE__*/function (_Error2) {
  createSuper._inherits(PendingSigningRequestError, _Error2);
  var _super2 = createSuper._createSuper(PendingSigningRequestError);
  function PendingSigningRequestError() {
    var _this2;
    createClass._classCallCheck(this, PendingSigningRequestError);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this2 = _super2.call.apply(_super2, [this].concat(args));
    wrapNativeSuper._defineProperty(createSuper._assertThisInitialized(_this2), "name", 'PendingSigningRequestError');
    wrapNativeSuper._defineProperty(createSuper._assertThisInitialized(_this2), "message", 'Clear or approve the pending wallet request and retry.');
    return _this2;
  }
  return createClass._createClass(PendingSigningRequestError);
}( /*#__PURE__*/wrapNativeSuper._wrapNativeSuper(Error));
var WalletConnectionErrorReason = /*#__PURE__*/function (WalletConnectionErrorReason) {
  WalletConnectionErrorReason["INCORRECT_CHAIN"] = "INCORRECT_CHAIN";
  WalletConnectionErrorReason["WRONG_ACCOUNT"] = "WRONG_ACCOUNT";
  WalletConnectionErrorReason["STALE_CONNECTION_REQUEST"] = "STALE_CONNECTION_REQUEST";
  return WalletConnectionErrorReason;
}({});
var WalletConnectionError = /*#__PURE__*/function (_Error3) {
  createSuper._inherits(WalletConnectionError, _Error3);
  var _super3 = createSuper._createSuper(WalletConnectionError);
  function WalletConnectionError(reason) {
    var _this3;
    createClass._classCallCheck(this, WalletConnectionError);
    _this3 = _super3.call(this, "Wallet connection failed due to ".concat(reason, " error"));
    wrapNativeSuper._defineProperty(createSuper._assertThisInitialized(_this3), "name", 'WalletConnectionError');
    _this3.reason = reason;
    return _this3;
  }
  return createClass._createClass(WalletConnectionError);
}( /*#__PURE__*/wrapNativeSuper._wrapNativeSuper(Error));
var UserRejectedError = /*#__PURE__*/function (_Error4) {
  createSuper._inherits(UserRejectedError, _Error4);
  var _super4 = createSuper._createSuper(UserRejectedError);
  function UserRejectedError() {
    var _this4;
    createClass._classCallCheck(this, UserRejectedError);
    for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      args[_key2] = arguments[_key2];
    }
    _this4 = _super4.call.apply(_super4, [this].concat(args));
    wrapNativeSuper._defineProperty(createSuper._assertThisInitialized(_this4), "name", 'UserRejectedError');
    wrapNativeSuper._defineProperty(createSuper._assertThisInitialized(_this4), "message", 'User rejected the request');
    return _this4;
  }
  return createClass._createClass(UserRejectedError);
}( /*#__PURE__*/wrapNativeSuper._wrapNativeSuper(Error));
var Wallet = /*#__PURE__*/createClass._createClass(function Wallet(address) {
  createClass._classCallCheck(this, Wallet);
  this.address = address;
});

exports.Profile = PublicationReport.Profile;
exports.PublicationReportReason = PublicationReport.PublicationReportReason;
exports.isMomokaPublicationId = Publication.isMomokaPublicationId;
exports.DataTransaction = Transactions.DataTransaction;
exports.MetaTransaction = Transactions.MetaTransaction;
exports.NativeTransaction = Transactions.NativeTransaction;
exports.ProtocolTransactionKinds = Transactions.ProtocolTransactionKinds;
exports.TransactionError = Transactions.TransactionError;
exports.TransactionErrorReason = Transactions.TransactionErrorReason;
exports.TransactionEvent = Transactions.TransactionEvent;
exports.TransactionKind = Transactions.TransactionKind;
exports.UnsignedTransaction = Transactions.UnsignedTransaction;
exports.InsufficientGasError = InsufficientGasError;
exports.NftContractType = NftContractType;
exports.PendingSigningRequestError = PendingSigningRequestError;
exports.ProfileReportReason = ProfileReportReason;
exports.UserRejectedError = UserRejectedError;
exports.Wallet = Wallet;
exports.WalletConnectionError = WalletConnectionError;
exports.WalletConnectionErrorReason = WalletConnectionErrorReason;
