// @lit-protocol/node-client/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.uint8arrayToString = exports.uint8arrayFromString = exports.blobToBase64String = exports.base64StringToBlob = exports.humanizeAccessControlConditions = exports.hashResourceIdForSigning = exports.zipAndEncryptString = exports.zipAndEncryptFiles = exports.verifyJwt = exports.encryptZip = exports.encryptString = exports.encryptFileAndZipWithMetadata = exports.encryptFile = exports.decryptZipFileWithMetadata = exports.decryptZip = exports.decryptString = exports.decryptFile = void 0;
const tslib_1 = require("tslib");
const misc_1 = require("@lit-protocol/misc");
const _NodeClient = require("./lib/node-client");
// ==================== Environment ====================
if ((0, misc_1.isNode)()) {
    (0, misc_1.log)('Oh hey you are running in Node.js!');
    const fetch = require('node-fetch');
    globalThis.fetch = fetch;
}
const NodeClient = _NodeClient.NodeClient;
if (!globalThis.NodeClient) {
    globalThis.NodeClient = NodeClient;
}
// ==================== Exports ====================
tslib_1.__exportStar(require("./lib/node-client"), exports);
var encryption_1 = require("@lit-protocol/encryption");
Object.defineProperty(exports, "decryptFile", { enumerable: true, get: function () { return encryption_1.decryptFile; } });
Object.defineProperty(exports, "decryptString", { enumerable: true, get: function () { return encryption_1.decryptString; } });
Object.defineProperty(exports, "decryptZip", { enumerable: true, get: function () { return encryption_1.decryptZip; } });
Object.defineProperty(exports, "decryptZipFileWithMetadata", { enumerable: true, get: function () { return encryption_1.decryptZipFileWithMetadata; } });
Object.defineProperty(exports, "encryptFile", { enumerable: true, get: function () { return encryption_1.encryptFile; } });
Object.defineProperty(exports, "encryptFileAndZipWithMetadata", { enumerable: true, get: function () { return encryption_1.encryptFileAndZipWithMetadata; } });
Object.defineProperty(exports, "encryptString", { enumerable: true, get: function () { return encryption_1.encryptString; } });
Object.defineProperty(exports, "encryptZip", { enumerable: true, get: function () { return encryption_1.encryptZip; } });
Object.defineProperty(exports, "verifyJwt", { enumerable: true, get: function () { return encryption_1.verifyJwt; } });
Object.defineProperty(exports, "zipAndEncryptFiles", { enumerable: true, get: function () { return encryption_1.zipAndEncryptFiles; } });
Object.defineProperty(exports, "zipAndEncryptString", { enumerable: true, get: function () { return encryption_1.zipAndEncryptString; } });
var access_control_conditions_1 = require("@lit-protocol/access-control-conditions");
Object.defineProperty(exports, "hashResourceIdForSigning", { enumerable: true, get: function () { return access_control_conditions_1.hashResourceIdForSigning; } });
Object.defineProperty(exports, "humanizeAccessControlConditions", { enumerable: true, get: function () { return access_control_conditions_1.humanizeAccessControlConditions; } });
var misc_browser_1 = require("@lit-protocol/misc-browser");
Object.defineProperty(exports, "base64StringToBlob", { enumerable: true, get: function () { return misc_browser_1.base64StringToBlob; } });
Object.defineProperty(exports, "blobToBase64String", { enumerable: true, get: function () { return misc_browser_1.blobToBase64String; } });
var uint8arrays_1 = require("@lit-protocol/uint8arrays");
Object.defineProperty(exports, "uint8arrayFromString", { enumerable: true, get: function () { return uint8arrays_1.uint8arrayFromString; } });
Object.defineProperty(exports, "uint8arrayToString", { enumerable: true, get: function () { return uint8arrays_1.uint8arrayToString; } });
// ----- autogen:polyfills:start  -----
// This field will be automatically injected into the ./dist/packages/<package-name>/index.js file 
// between the autogen:polyfills:start/end comments

try {
    globalThis.crypto = require('crypto').webcrypto;
} catch (e) {
    // swallow
}
// ----- autogen:polyfills:end  -----
